File: Practical/practices/pract3.c
#include<stdbool.h>

void bfs(int graph[7][7],int startNode, int endNode){
    bool visited[7];
    int queue[7];
    int front =0,  rear=0;

    for(int i=0;i<7;i++){
        visited[i] = false;
    }

    queue[rear++] = startNode;
    visited[startNode] = true;

    while(front != rear){
        int currentNode = queue[front++];
        if(currentNode != endNode){
            printf("->");
        }else{
            printf("Target Not Found ");
            return;
        }
        for(int i=0;i<7;i++){
                if(!visited[i] && graph[currentNode][i] == 1){
                    visited[i] = true;
                    queue[rear++] = i;
                }
        }
    }
    printf("Target  Found !!");

}

File: Practical/python/pract3.py
from collections import deque

def bfs(graph, start_node, target_node):
    n = len(graph)
    visited = [False] * n
    queue = deque()

    visited[start_node] = True
    queue.append(start_node)

    print(f"Breadth-First Traversal [starting from node {chr(65 + start_node)}]:")
    
    while queue:
        current_node = queue.popleft()
        print(f"{chr(65 + current_node)}", end="")
        
        if current_node != target_node:
            print("->", end="")
        else:
            print("\nTarget node found.")
            return
        
        for i in range(n):
            if graph[current_node][i] == 1 and not visited[i]:
                visited[i] = True
                queue.append(i)
    
    print("Target node not found.")

if __name__ == "__main__":
    n = 7
    adjacency_matrix = [
        [0, 1, 0, 1, 0, 0, 0],
        [0, 0, 1, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0, 1],
        [0, 0, 0, 0, 0, 1, 0],
        [1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 1, 0, 0]
    ]
    start_node = 0
    target_node = 4

    bfs(adjacency_matrix, start_node, target_node)
